impl/src/main/java/org/jboss/weld/util/collections/AbstractMultimap.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2014, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.util.collections;
+
+import static org.jboss.weld.util.Preconditions.checkArgumentNotNull;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+
+import org.jboss.weld.util.Supplier;
+
+/**
+ * An abstract {@link Multimap} backed by a {@link Map}.
+ *
+ * @author Martin Kouba
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ * @param <C> The collection of values type
+ */
+abstract class AbstractMultimap<K, V, C extends Collection<V>> implements Multimap<K, V>, Serializable {
+
+    private static final long serialVersionUID = -8363450390652782067L;
+
+    protected final Supplier<C> supplier;
+
+    private final Map<K, C> map;
+
+    /**
+     *
+     * @param mapSupplier
+     * @param collectionSupplier
+     * @param multimap
+     */
+    protected AbstractMultimap(Supplier<Map<K, C>> mapSupplier, Supplier<C> collectionSupplier, Multimap<K, V> multimap) {
+        checkArgumentNotNull(mapSupplier, "mapSupplier");
+        checkArgumentNotNull(collectionSupplier, "collectionSupplier");
+        this.supplier = collectionSupplier;
+        this.map = mapSupplier.get();
+        if (multimap != null) {
+            for (Entry<K, Collection<V>> entry : multimap.entrySet()) {
+                C values = supplier.get();
+                values.addAll(entry.getValue());
+                putAll(entry.getKey(), values);
+            }
+        }
+    }
+
+    @Override
+    public int size() {
+        return map.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return map.isEmpty();
+    }
+
+    @Override
+    public C get(K key) {
+        C value = map.get(key);
+        if (value == null) {
+            value = supplier.get();
+            C previousValue = map.put(key, value);
```

Comment:
As previously discussed this is not thread safe. We cannot use `Map.computeIfAbsent()` or `Map.putIfAbsent()` introduced in JDK8 so I propose to add `ConcurrentMap<K, C> concurrentMap` field and corresponding param to the constructor. It would be illegal to provide both suppliers.

The `get()` method could have two branches. One for regular map:

```
if (map.get(key) == null) {
     V newValue = supplier.get(key);
     if (newValue != null)
         map.put(key, newValue);
 }
```
And second branch for `ConcurrentMap` - the same as the current version but using `putIfAbsent()` instead of `put()`.
Of course, we would have to modify `SetMultimap` and `ListMultimap`.